### ISSUE TITLE
Need ODM SDF Syntax to Cover Thing / Interface Object / Trait Version Mappings

### ISSUE THREAD

**Issue by gerickson** [one-data-model/language#60]
In Weave Data Language (WDL), interfaces are equivalent to ODM "things" and describe an abstract, contractual publish/subscribe interface for generic ecosystem interactions. Over time, the definition of the interface may evolve and its version incremented as those changes occur. To the extent that changes in the interface need to be bound to specific versions of one or more underlying traits (objects), WDL provides a construct in the form of the _version_map_ decoration to specify those dependencies.

To be expressed in ODM SDF losslessly, a syntax construct for representing this version mapping, when present, is required.

### Relevant Example Schema Input Files

* [weave/iface/lock/voice_assistant_lock_iface.proto](https://github.com/openweave/openweave-schema-vendor-common/blob/master/weave/iface/lock/voice_assistant_lock_iface.proto)

### Missing or Unmapped WDL-to-ODM SDF Syntax

* **version_map**
    - This is an array of dictionaries specifying which specific versions of the traits comprising the interface are depended upon for a particular interface version. The dictionaries contain two keys: _parent_version_ and _dependent_version_list_. Note that not all traits specified in the interface need be specified in the version map; they need only be specified if there are specific version requirements.
    - **dependent_version_list**
        - This is an array of dictionaries specifying the trait and version thereof that is to be bound for a particular interface version. The dictionaries contain two keys: _name_ and _version_.
        - **name**
The name of the trait on which the specific _parent_version_ of this interface depends.
        - **version**
The version of the named trait on which the specific _parent_version_ of this interface depends.
    - **parent_version**
        - The specific version of the interface for which trait dependencies are being specified.

#### Example Input WDL

```
message BasicLockIface {
    option (wdl.message_type) = IFACE;

    option (wdl.iface) = {
        stability: ALPHA,
        vendor_id: 0x0000,
        version: 1
  };

    weave.trait.security.BoltLockTrait          lock             = 1;

    weave.trait.security.BoltLockSettingsTrait  lock_settings    = 2;

    weave.trait.auth.ApplicationKeysTrait       application_keys = 3;
}

message VoiceAssistantLockIface {
    option (wdl.message_type) = IFACE;

    option (wdl.iface) = {
        stability: ALPHA,
        vendor_id: NEST,
        version: 1,
        version_map: [
            {
                parent_version: 1, dependent_version_list: [
                    { name: "weave.trait.security.BoltLockTrait", version: 2 }
                ]
            }
        ]
    };

    weave.trait.security.BoltLockTrait bolt_lock = 1;
}
```

**Comment by mjkoster** [comment]
This should be resolved using the same mechanism as #40 so it seems to be a duplicate

**Comment by gerickson** [comment]
Agreed that we can hopefully cover the requirement with the same approach. Not sure why Google didn't do so originally.

**Comment by asoloway64** [comment]
F2F4: OneDM Language and OneDM data models shall both be versioned (i.e. a "release tag" in github). When the models are "released", both the language and the models get the same "release tag". Changes to the language shall not require updates to the existing models. The language can be released without re-releasing the models.

**Comment by asoloway64** [comment]
F2F4: Version negotiation would be a function of protocol binding and is not represented in the schema. The preferred way to adapt to changes in the data model is through feature and object discovery.
